dask/blockwise.py

```diff
@@ -318,6 +336,32 @@ def __dask_distributed_pack__(self, client):
         # All blockwise tasks will depend on the futures in `indices`
         global_dependencies = tuple(stringify(f.key) for f in indices_unpacked_futures)
 
+        # Check if we need to serialize the elements of `io_deps`
+        # with `pickle` (i.e. something in the required inputs is
+        # not msgpack serializable)
+        if msgpack:
+            for _, input_map in self.io_deps.items():
+                check = True
+                for k in input_map:
+                    # Test msgpack on the very first element.
+                    # If this fails, we will pickle everything.
+                    # If this succeeds, we will assume all other
+                    # elelments are also msgpack serializable.
+                    #
+                    # The risk here is that only a subset of
+                    # elements requires pickling, and that subset
+                    # does not include the first element.  For
+                    # now, it is the responsibility of the
+                    # Blockwise layer to pickling/unpickle
+                    # offending elements in cases like this.
+                    if check:
+                        try:
+                            msgpack.packb(input_map[k])
+                            break  # No need to pickle - Bail
+                        except TypeError:
+                            check = False
+                    input_map[k] = pickle.dumps(input_map[k])
```

Comment:
@jrbourbeau - I think this PR is ready.  However, I am still not sure of the best way to check if the `io_deps` need to be pickled.  A relatively reliable approach (used in the current version of this PR) is to check if the first element can be serialized with `msgpack`, but this is far from perfect (and requires a try/except).

Note that this pickling is currently only needed for specific options in `read_parquet(.., engine="pyarrow-dataset")` and for `read_csv` (all other cases are msgpack serializable).  This same problem/question was also mentioned in [another comment](https://github.com/dask/dask/pull/7042/files#r568102871)).